### PR TITLE
SAMZA-2018 : State restore improvements

### PR DESCRIPTION
**Pull request by rmatharu**
This PR makes the following changes: 

* Consumer consolidation to ensure 1 storeConsumer per system, earlier it was 1 consumer per SSP per store.
* Refactoring stores to use ContainerStorageManager with parallelization for restoration, and serial execution of sysConsumers start, stop, register, etc.

### CODE REVIEW

**Comment by prateekm** [comment]
@vjagadish1989 please take a look.

**Comment by rmatharu** [comment]
Applied the comments -- removed unused variables, simplified expressions for storeConsumers, etc.

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  /** Callable for performing the restoreStores on a taskStorage manager and emitting task-restoration metric.
+   *
+   */
+  private class TaskRestoreRunnable implements Callable {
+
+    private TaskInstance taskInstance;
+    private TaskStorageManager taskStorageManager;
+    private SamzaContainerMetrics samzaContainerMetrics;
+
+    public TaskRestoreRunnable(SamzaContainerMetrics samzaContainerMetrics, TaskInstance taskInstance,
```
[the remark this answers is not in the document]
Comment (rmatharu) [reply samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: yes Callable allows easily propagating any exceptions occurring during restores.
Runnable doesnt.

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    this.taskStorageManagers.entrySet().forEach(task -> {
+        futureList.add(
+            executorService.submit(new TaskRestoreRunnable(this.samzaContainerMetrics, task.getKey(), task.getValue())));
+      });
+
+    // loop-over the future list to wait for each thread to finish, catch any exceptions during restore and throw
+    // as samza exceptions
+    for (Future future : futureList) {
+      try {
+        future.get();
```
[the remark this answers is not in the document]
Comment (rmatharu) [reply samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: We have to wait till the entire dataset is loaded in, without that the container cannot make progress anyways.

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala [hunk 3]
```diff
@@ -502,6 +502,31 @@ object SamzaContainer extends Logging {
 
     val timerExecutor = Executors.newSingleThreadScheduledExecutor
 
+    // Create a map of storeName to store's SystemName
+    val storeSystems: Map[String, String] = changeLogSystemStreams.
```
[the remark this answers is not in the document]
Comment (rmatharu) [reply samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala]: done

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala [hunk 4]
```diff
@@ -502,6 +502,23 @@ object SamzaContainer extends Logging {
 
     val timerExecutor = Executors.newSingleThreadScheduledExecutor
 
+    // Create one SystemConsumer for each storeSystem in use
```
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala]: 505 and 506 are redundant comments? maybe nuke one, retain another

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ *  a) performing all container-level serializable actions for restore, initializing and shutting down
+ *  taskStorage managers, starting and stopping consumers, etc.
+ *
+ *  b) performing parallelized taskStorageManager restores.
+ *
+ */
+public class ContainerStorageManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ContainerStorageManager.class);
+  private final Map<TaskInstance, TaskStorageManager> taskStorageManagers;
```
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: group statics and non-statics

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+import org.apache.samza.container.SamzaContainerMetrics;
+import org.apache.samza.container.TaskInstance;
+import org.apache.samza.metrics.Gauge;
+import org.apache.samza.system.SystemConsumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ *  ContainerStorageManager is a per-container abstraction that instruments
```
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: s/instruments/manages

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ *  the restore of per-task partitions.
+ *
+ *  It is responsible for
+ *  a) performing all container-level serializable actions for restore, initializing and shutting down
```
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: Make docs clearer. for eg: remove "serializable actions". s/performing parallelized taskstorageManager restores/restoring individual stores in parallel

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java (hunk 6 above)
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: remove words like "abstraction" from java-docs

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  // Mapping of from storeSystemNames to SystemConsumers
+  private final Map<String, SystemConsumer> systemConsumers;
+
+  // Size of thread-pool to be used for parallel restores
+  private final int parallelRestoreThreadPoolSize;
+
+  // Naming convention to be used for restore threads
+  private static final String RESTORE_THREAD_NAME = "Samza Restore Thread-%d";
+
+  public ContainerStorageManager(Map<TaskInstance, TaskStorageManager> taskStorageManagers,
```
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: for a future PR, we should evaluate re-purposing this as a "RestoreManager" that only handles restores.. thereby stripping restore logic away from the TSM. The TSM wouldn't need access to `SystemConsumer` instances as well

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    this.taskStorageManagers.values().forEach(taskStorageManager -> taskStorageManager.init());
+
+    // Start consumers
+    this.systemConsumers.values().forEach(systemConsumer -> systemConsumer.start());
+
+    // Create a thread pool for parallel restores
+    ExecutorService executorService = Executors.newFixedThreadPool(this.parallelRestoreThreadPoolSize,
+        new ThreadFactoryBuilder().setNameFormat(RESTORE_THREAD_NAME).build());
+
+    List<Future> futureList = new ArrayList<>(this.taskStorageManagers.entrySet().size());
```
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: taskRestoreFutures

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java (hunk 8 above)
Comment (rmatharu) [reply samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: yup, agreed.

**Comment by rmatharu** [comment]
Done, applied all comments.

**Pull request closed without merging**